concepts.tex

```diff
@@ -803,6 +564,321 @@
+names of concepts that traditionally describe features of types to describe families of object types (See
+the rationale in Appendix~\ref{decomposition}).}
+
+\rSec2[concepts.lib.object.destructible]{Concept Destructible}
+\ednote{Remove table [destructible] in [utility.arg.requirements]. Replace
+references to [destructible] with references to
+[concepts.lib.\oldtxt{corelang}\newtxt{object}.destructible].}
+
+\pnum
+\newtxt{The \tcode{Destructible} concept is the base of the hierarchy of object concepts.
+It describes properties that all such object types have in common.}
+
+\indexlibrary{\idxcode{Destructible}}%
+\begin{itemdecl}
+template <class T>
+concept bool Destructible@\newtxt{() \{}\oldtxt{ =}@
+  @\newtxt{return}@ requires (T t@\newtxt{, T* p}@) {
+    { t.@$\sim$@T() } noexcept;
+    @\newtxt{\{ \&t \} -> Same<T*>;}@
+    @\newtxt{delete p;}@
+    @\newtxt{delete[] p;}@
+  };
+@\newtxt{\}}@
+\end{itemdecl}
+
+\begin{itemdescr}
+\ednote{\oldtxt{REVIEW: The expression \tcode{t.$\sim$T()} comes straight from the requirements table for the
+\tcode{Destructible} concept in C++14. However, \tcode{t.$\sim$T()} is not valid if \tcode{T} is a
+reference or array type. We need to decide whether we want \tcode{Destructible<int\&>} and
+\tcode{Destructible<int[42]>} to be true or false.}}
+
+\pnum
+\oldtxt{A}\newtxt{Given an object \tcode{t} of type \tcode{T} and pointer \tcode{p}
+of type \tcode{T*}, the} type \tcode{T} models \tcode{Destructible} if and only if
+
+\begin{itemize}
+\item After \newtxt{evaluating} the expression \tcode{\oldtxt{u}\newtxt{t}.$\sim$T()},
+\newtxt{\tcode{delete p}, or \tcode{delete[] p},} all resources owned by
+\oldtxt{\tcode{u}}\newtxt{the denoted object(s)} are reclaimed.
+\item \newtxt{\tcode{\&t == addressof(t)}.}
+\end{itemize}
+\end{itemdescr}
+
+\rSec2[concepts.lib.object.constructible]{Concept Constructible}
+
+\pnum
+\newtxt{The \tcode{Constructible} concept is used to constrain the type of a
+variable to be either an object type constructible from a given set of argument
+types, or a reference type that can be bound to those arguments.}
+
+\indexlibrary{\idxcode{Constructible}}%
+\begin{itemdecl}
+template <class T, class ...Args>
+concept bool Constructible@\newtxt{() \{}\oldtxt{ =}@
+  @\newtxt{return}@ requires (@\oldtxt{ }@Args@\newtxt{\&\&}@...args@\oldtxt{ }@) {
+    T{forward<Args>(args)...};
+  }@\oldtxt{;}\newtxt{ \&\&}@
+  @\newtxt{(is_reference<T>::value ||}@
+    @\newtxt{(Destructible<T>() \&\& requires (Args\&\&...args) \{}@
+      @\newtxt{new T\{forward<Args>(args)...\};}@
+    @\newtxt{\}));}@
+@\newtxt{\}}@
+\end{itemdecl}
+
+\ednote{\newtxt{Requiring \tcode{T\{forward<Args>(args)...\}} to be equality preserving
+implies that construction from equal arguments must result in equal objects, i.e., the
+value of a \tcode{Constructible} object must be totally determined by the constructor
+arguments.}}
```

Comment:
I think I'm ready to admit now that you are right about this. That property goes too far in that it enforces a particular style of programming which is antithetical to the multi-paradigm design of C++. I've also just realized that `new T{...}` is inherently not equality preserving and needs an exception. I'll push a patch.